Classical control

daxfohl opened the pull request:
Sits on top of #4627

Creates `ConditionalOperation` class and executes operations conditionally upon the classical bits. Most of this is done in commit https://github.com/quantumlib/Cirq/pull/4631/commits/06883d5d3d767eb08fe77bc0e4a0cf6ae7e0d64c.

Reimplements quantum teleportation example based off this class.

Parts 8, 9, 10 of https://tinyurl.com/cirq-feedforward.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,155 @@
+        }
+
+    def _act_on_(self, args: 'cirq.ActOnArgs') -> bool:
+        def not_zero(measurement):
```
Actually I'm thinking `_controls` should be `List[Tuple[MeasurementKey, int], ...]` instead, allowing users to specify the index to control on. This would also make compatibility with QASM easier.

95-martin-orion replied:
[No change required] This function is kind of unnecessary in this PR, given that it's a one-liner, but I can see it becoming worthwhile once we start adding support for non-1 conditions.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,161 @@
+            [protocols.with_measurement_key_mapping(k, key_map) for k in self._controls],
+        )
+
+    def _control_keys_(self) -> Tuple[value.MeasurementKey, ...]:
```
In parity with _measurement_keys_, should add control keys from the suboperation here too. Also note this protocol needs added to TaggedOperation. Aside, should add logic to ControlledOperation constructor to fail if it wraps a ConditionalOp or measurement.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,161 @@
+    ):
+        controls = tuple(value.MeasurementKey(k) if isinstance(k, str) else k for k in controls)
+        if isinstance(sub_operation, ConditionalOperation):
+            # Auto-flatten nested controlled operations.
```
This idea was copied from ControlledOperaton and TaggedOperation. But I'm not so sure it's the right thing to do here. For instance, consider `ConditionalOp(TaggedOp(ConditionalOp))`. Should this expose the inner controls or only the outer? How about `ConditionalOp(TaggedOp(Subcircuit(ConditionalOp)))`? These should be exposed in the `control_keys` protocol I assume, but do they need to be in `op.controls`? Or should `op.controls` only show the top level controls? I could see this either way.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,161 @@
+        return self._controls
+
+    @property
+    def sub_operation(self):
```
Does this even need to be public? It could be private, then the "number of wrappers" problem is less troublesome. We can just design around the best way to make `op.with_tags().with_conditions().with_tags().without_conditions()....` behave the way we want. Plausibly even the `ConditionalOperation` class itself could be private.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,153 @@
+    def __init__(
+        self,
+        sub_operation: 'cirq.Operation',
+        keys: Union[str, 'cirq.MeasurementKey', Sequence[Union[str, 'cirq.MeasurementKey']]],
```
varargs

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,153 @@
+        self._sub_operation = sub_operation
+
+    @property
+    def conditions(self) -> FrozenSet['cirq.MeasurementKey']:
```
Consider returning a set of a new custom `Condition` interface here. In the future, we'll have more complicated conditions. Make this fairly minimal. For now, there should be `Condition.keys()`, and possibly that's all.

daxfohl replied:
Okay, so https://github.com/daxfohl/Cirq/compare/scope3...daxfohl:sympy?expand=1 shows roughly what the `Condition` will look like with the sympy conditions. A few options here.

1. We could integrate all that into this PR. That's a fairly large increase in scope (we don't even have solid requirements on that yet) and I'd probably not want to.
2. We could just create a skeleton `Condition` and use that here. I'm a bit reluctant on this because even a skeleton Condition is a fair amount of extra work (serialization, testing, etc) and it may not even match the final sympy code.
3. We could leave this as-is and just make it a breaking change when/if sympy or other condition expressions get integrated.
4. We could rename this to `control_keys` and then deprecate it (or not) once we have a proper Condition class for a proper `conditions` property.

I somewhat lean toward (4) right now. One caveat of this is that `op.control_keys()` will be different than `protocols.control_keys(op)`, once the scoping PR gets merged. But we've talked about this before and it seems we're okay with that. The other caveat is that I do still think we should keep `with_conditions` and `unconditionally` as-is, and the name mismatch could be confusing. But that would only be short-term until we get a proper Condition class and method, so I think it's okay.

daxfohl replied:
Even better, I decided just to delete it. We do have a `control_keys` protocol, so we can use that instead to get control keys. Then we can add the `conditions` method once it makes sense to do so.

daxfohl commented on cirq-core/cirq/ops/raw_types.py:
```diff
@@ -602,6 +603,20 @@ def _commutes_(
+        return frozenset()
+
+    def with_conditions(
+        self, keys: Union[str, value.MeasurementKey, Sequence[Union[str, value.MeasurementKey]]]
```
varargs

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,153 @@
+    def _value_equality_values_(self):
+        return (frozenset(self._control_keys), self._sub_operation)
+
+    def __str__(self) -> str:
```
Make a simpler str representation

95-martin-orion commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,143 @@
+
+
+@value.value_equality
+class ConditionalOperation(raw_types.Operation):
```
[Bikeshed] We should vote on this name at the Cirq sync - my intuition is that "conditional" may be too semantically similar to "controlled" for users to consistently remember which is which.

Might end up being a moot point if we push for merging conditional / controlled / tagged operations into the base Operation class, but we should be clear in the meantime.

daxfohl replied:
FWIW I actually kind of like the wrapper classes now. Or maybe I've just gotten used to them.

95-martin-orion replied:
(unresolving so we remember to discuss at Cirq sync)

95-martin-orion commented on cirq-core/cirq/ops/conditional_operation_test.py:
```diff
@@ -0,0 +1,353 @@
+"""
+    )
+
+
```
There are some negative tests I'd like to see added here:
- Condition on a measurement key that has not been measured yet
- Inside a subcircuit, condition on a measurement key outside that subcircuit
- Tests around conditional measurement (I assume there's potential for misbehavior if you do `cirq.measure(q0, key='m').with_conditions('a')`, especially if `'a'` is zero)

If you can think of anything of tests of this type (i.e. testing unsupported or invalid behaviors), please add them as well.

daxfohl replied:
Sure. I can tell you what the behaviors will be:

> Condition on a measurement key that has not been measured yet
Runtime error, key not found. I'll add a better error message.

> Inside a subcircuit, condition on a measurement key outside that subcircuit
Even though it's not officially supported, this will magically hook up iff there are no repetition ids (runtime error otherwise). I propose just letting this behavior be "undefined" for now. All that is fully fleshed out and tested in the scope PR that's next in line.

> Tests around conditional measurement
This would behave as expected (like the gate was there or it was not), but I'll add a test. Though that does bring up whether to even allow this. My gut feeling is no, at least initially. Thoughts?

daxfohl replied:
I also think we should prohibit any of these as ControlledOperation subops.

95-martin-orion replied:
Agreed on prohibiting `ConditionalOperation`s as `ControlledOperation` subops - if possible, they should be prohibited by the same rules that prevents controlled measurement.

I vote for allowing conditional measurement, as it's necessary for some useful constructions:
```python
circuit_a = Circuit(
    cirq.measure(q0, key='k'),
    # if k, run subcircuit with self-contained measurements
    CircuitOperation(FrozenCircuit(
        cirq.measure(q1, key='m'),
        cirq.X(q1).with_conditions('m')
    )).with_conditions('k')
)

circuit_b = Circuit(
    cirq.measure(q0, key='k'),
    # if k, measure q1; else measure q2
    cirq.measure(q1, key='m').with_conditions('k')
    cirq.measure(q2, key='m').with_conditions('k', invert_mask=[True])
)
```

daxfohl replied:
Yeah, the "at least not initially" was intentional. The thing that concerns me is
```python
Circuit(
    cirq.measure(q0, key='k'),
    cirq.measure(q1, key='m').with_conditions('k'),
    cirq.X(q1).with_conditions('m')
)
```
, which would be a valid circuit (under the existing logic) but throw a runtime error.

I'd want to make sure we have a design that prevents such circuits from being constructed first. In you first example, maybe we say `cirq.measure(q1, key='m', local_key=True)` or something, and for the second one we make a new `IfElseOperation(op_if, op_else, condition)` that ensures both ops have the same exposed measurement keys.

Or maybe I am overthinking it and we just let users figure it out themselves.

95-martin-orion replied:
+1 to not allowing this in this PR (incl. tests to verify this), and revisiting later.

This essentially boils down to "reference to possibly-uninitialized variable". Coming from C++ and Python where this is permitted (albeit with a warning in C++), I'm in favor of letting users handle this; however, given that there are languages like Java which do not allow this, I can see either choice being valid.

daxfohl replied:
FWIW we don't currently have a check against controlled measurements. I will add one. There are probably other things that shouldn't be controlled (reset gate?) but I'll leave that for later.

daxfohl replied:
Had to undo the check against controlled measurements. That broke consistent protocols test, and was outside the scope of this PR anyway. We can revisit that later. Added the remaining tests and required functionality to conditional_op though.

95-martin-orion commented on cirq-core/cirq/ops/conditional_operation_test.py:
```diff
@@ -0,0 +1,353 @@
+
+
+@pytest.mark.parametrize('sim', ALL_SIMULATORS)
+def test_subcircuit_key_unset(sim):
```
Would tests of `CircuitOperation(...).with_conditions(...)` be interesting?

daxfohl replied:
Not really. A subop is a subop. But I can add one.

daxfohl replied:
Actually, I think that would just burn CPU unnecessarily. More useful would be the test that prohibits circuitoperations with measurements as suboperations. (If we decide to prohibit them).

95-martin-orion commented on cirq-core/cirq/protocols/json_test_data/ConditionalOperation.repr:
```diff
@@ -0,0 +1 @@
+cirq.ConditionalOperation(cirq.Y.on(cirq.NamedQubit('target')), [cirq.LineQubit(0), cirq.LineQubit(1)])
```
This doesn't seem like a valid repr for this type - `cirq.LineQubit(0)` is not a `MeasurementKey`.

95-martin-orion commented on cirq-core/cirq/sim/clifford/stabilizer_sampler.py:
```diff
@@ -53,7 +53,7 @@ def run_sweep(
     def _run(self, circuit: circuits.AbstractCircuit, repetitions: int) -> Dict[str, np.ndarray]:
 
-        measurements: Dict[str, List[int]] = {
+        measurements: Dict[str, List[np.ndarray]] = {
```
Why does this change?

daxfohl replied:
This seems to have been incorrectly typed all along. It only worked because `log_of_measurement_results` values were `Any`. I had to change the latter so that the condition logic would work, and that highlighted this error here.

95-martin-orion reviewed (changes requested):
Top-level comment: could you open an issue on the [qsim repo](https://github.com/quantumlib/qsim) to add support for classically-controlled operations?

I mostly left the circuit diagram stuff alone here since it was covered in previous PRs. If there are significant changes here I should look at, please let me know.

daxfohl commented on cirq-core/cirq/ops/moment.py:
```diff
@@ -220,7 +220,7 @@ def without_operations_touching(self, qubits: Iterable['cirq.Qid']) -> 'cirq.Mom
         return Moment(
             protocols.with_measurement_key_mapping(op, key_map)
-            if protocols.is_measurement(op)
+            if protocols.is_measurement(op) or protocols.control_keys(op)
```
I was actually thinking something like `interacts_with_registers` or something. This `or` check occurs three times in this PR, so would be good to consolidate. Thoughts?

95-martin-orion replied:
I dont' think "registers" is defined in Cirq, and it's a bit ambiguous besides - a register could be classical or quantum. Aside from that, I'm in support of adding syntactic-sugar methods to tidy this up.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,143 @@
+        def not_zero(measurement):
+            return any(i != 0 for i in measurement)
+
+        measurements = [args.log_of_measurement_results[str(key)] for key in self._control_keys]
```
Probably backwards compatibility. Or maybe it just never got noticed between Smit and me. I'd say if we do change it, maybe it's better to make a new field like `registers`, especially if we might consider classical operations like xor, concat, etc on them in the future. Then we deprecate `log_of_measurement_results`.

daxfohl commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,143 @@
+    ):
+        keys = tuple(value.MeasurementKey(k) if isinstance(k, str) else k for k in conditions)
+        self._control_keys = cast(Tuple[value.MeasurementKey, ...], keys)
+        self._sub_operation = sub_operation
```
We *could* but I'd suggest not, as there's not really a reason to. TaggedOperation *has* to, since tags don't propagate from the suboperation by design. I think for ConditionalOperation it's cleaner just to leave it as-is.

95-martin-orion replied:
I disagree about there not being a reason to flatten. Nested `ConditonalOperation`s cause bloat in string-output methods (e.g. `CO(keys=a, sub_op=CO(keys=b, sub_op=op))` vs `CO(keys=[a, b], sub_op=op)`) and are more expensive to operate on than flat `ConditionalOperation`s (e.g. when using the recursive `unconditionally` method).

daxfohl replied:
Done, but note that ControlledOperation is the one that does this in the constructor. TaggedOperation does this in the with_tags function. So TaggedOperation(TaggedOperation(op), 'a'), 'b') would have only 'b' as a tag. I used the ControlledOperation approach. We should follow up and make this consistent.

daxfohl commented on cirq-core/cirq/circuits/circuit_test.py:
```diff
@@ -247,192 +225,19 @@ def test_append_single():
     assert len(c) == 1
 
 
-def test_control_key_diagram():
```
I moved them over to controlled_operation_tests. This file is already pretty huge, and I figured it made more sense for them to be there anyway.

daxfohl commented on cirq-core/cirq/ops/raw_types.py:
```diff
@@ -777,6 +788,12 @@ def _equal_up_to_global_phase_(
         return protocols.equal_up_to_global_phase(self.sub_operation, other, atol=atol)
 
+    def unconditionally(self) -> Operation:
+        return TaggedOperation(self.sub_operation.unconditionally(), *self.tags)
```
This isn't right. The tags should be removed, *unless* `self.sub_operation is self.sub_operation.unconditionally()`. Basically if it leaves the suboperation unchanged, return self, otherwise don't retag the returned operation.

daxfohl commented on cirq-core/cirq/ops/conditional_operation_test.py:
```diff
@@ -0,0 +1,353 @@
+ALL_SIMULATORS = (
+    cirq.Simulator(),
+    cirq.DensityMatrixSimulator(),
+    cirq.CliffordSimulator(),
```
Add MPS

95-martin-orion reviewed (commented):
Changes look good. I think the only remaining items are the discussion on what to call this operation (to be covered at Cirq sync today) and the "controlled CircuitOp which contains a measurement" test case. After these are done, we can merge.

daxfohl commented:
It looks like `IfOperation` won. Are we keeping `with_conditions()` and `unconditionally()`?

95-martin-orion commented:
> It looks like `IfOperation` won. Are we keeping `with_conditions()` and `unconditionally()`?

We still need to confirm with the internal error-correction team on the name, but either way I think those methods are still reasonable names.

dabacon commented on cirq-core/cirq/ops/conditional_operation.py:
```diff
@@ -0,0 +1,154 @@
+    def __init__(
+        self,
+        sub_operation: 'cirq.Operation',
+        conditions: Sequence[Union[str, 'cirq.MeasurementKey']],
```
what is the meaning of multiple conditions and measurement keys that correspond to more than one value?

daxfohl replied:
I'll add some documentation here.

Multiple conditions are treated as `and`, so chaining `op.with_conditions('a', 'b').with_conditions('c', 'd')` is equivalent to `op.with_conditions('a', 'b', 'c', 'd')`.

If a measurement was of multiple qubits, all zeros is considered False, anything else is considered True.

There's a subsequent PR that will allow sympy expressions to be conditions, like `'mkey1 | mkey2'`, or `'mkey1 > 20'`. That will enable `or` conditions, and measurements-with-multiple-qubits to be used as ints,

95-martin-orion commented:
Between the Cirq sync and internal review, `ClassicallyControlledOperation` appears to be the preferred name for this. Let's update to use that name and get this merged.

daxfohl commented:
@95-martin-orion I went ahead and updated the function names to `with(/out)_classical_controls` to match. LMK if you prefer reverting to `with_conditions`.

I added tests for subcircuits with control keys based on external measurements as well. They *work*, but I wouldn't say they're supported yet. As you can see in the tests, they require special handling during construction, since the control_keys protocol doesn't yet expose unbound control keys of subcircuits, so the logic in append doesn't detect the key conflict and pushes that subcircuit to the first moment by default. That is all taken care of in the scope PR.

Also I promised Dave some pydocs above. I'll add those.

daxfohl commented:
Pydocs, added. Ready for review.

95-martin-orion reviewed (approved):
Changes LGTM. Pinging @dabacon for his input on the docstrings, otherwise this is ready to go.

95-martin-orion commented:
Going ahead with merge to keep these moving.
